### ISSUE TITLE
Metabase doesn't sync in Australian Timezone

### ISSUE THREAD

**Issue by lindsay-stevens** [metabase/metabase#5869]
## Context

- Your browser and the version: Chrome 60.0
- Your operating system: Centos 6.9 (chrome client on Windows 7)
- Your databases: Postgres 9.6
- Metabase version: v0.26.0 RC1
- Metabase hosting environment: Linux/Centos 6
- Metabase internal database: Postgres 9.6

## Issue

The database sync feature for Postgres seems to barf when the system timezone is "AEST" (Australia/Sydney). Maybe because it's too long, since it works OK with "UTC". This problem wasn't occuring in my previous Metabase version, v0.22.

## Traceback

```bash
09-01 13:07:30 DEBUG metabase.middleware :: POST /api/database/2/sync_schema 200 (7 ms) (1 DB calls)
09-01 13:07:30 ERROR sync.util :: Error running sync step: Invalid format: "2017-09-01 13:07:30.784 AEST" is malformed at "AEST"
("driver$make_current_db_time_fn$fn__22930.invoke(driver.clj:215)"
 "driver$fn__22838$G__22690__22845.invoke(driver.clj:33)"
 "sync.sync_metadata.sync_timezone$fn__34467$sync_timezone_BANG___34472$fn__34473.invoke(sync_timezone.clj:18)"
 "sync.sync_metadata.sync_timezone$fn__34467$sync_timezone_BANG___34472.invoke(sync_timezone.clj:12)"
 "sync.sync_metadata$fn__34707$sync_db_metadata_BANG___34712$fn__34713$fn__34714.invoke(sync_metadata.clj:24)"
 "sync.util$do_with_error_handling.invokeStatic(util.clj:115)"
 "sync.util$do_with_error_handling.invoke(util.clj:110)"
 "sync.util$do_with_error_handling.invokeStatic(util.clj:113)"
 "sync.util$do_with_error_handling.invoke(util.clj:110)"
 "driver$fn__22934.invokeStatic(driver.clj:226)"
 "driver$fn__22934.invoke(driver.clj:226)"
 "driver$fn__22903$G__22684__22912.invoke(driver.clj:33)"
 "sync.util$sync_in_context$fn__32332.invoke(util.clj:106)"
 "sync.util$with_db_logging_disabled$fn__32329.invoke(util.clj:98)"
 "sync.util$with_start_and_finish_logging$fn__32324.invoke(util.clj:87)"
 "sync.util$with_sync_events$fn__32321.invoke(util.clj:70)"
 "sync.util$with_duplicate_ops_prevented$fn__32312.invoke(util.clj:49)"
 "sync.util$do_sync_operation.invokeStatic(util.clj:133)"
 "sync.util$do_sync_operation.invoke(util.clj:130)"
 "sync.sync_metadata$fn__34707$sync_db_metadata_BANG___34712$fn__34713.invoke(sync_metadata.clj:23)"
 "sync.sync_metadata$fn__34707$sync_db_metadata_BANG___34712.invoke(sync_metadata.clj:20)"
 "api.database$fn__36954$fn__36955$fn__36956.invoke(database.clj:495)")
```

## Repro

I'll work backwards. To try and solve this I tried:

- Setting the Metabase->General settings to "database default" w/ restart
- Changing the postgresql.conf timezone settings to UTC (both log timezone and client timezone) w/ restart

But it didn't work until I set the system timezone to UTC, restarted Metabase (Centos 6), and re-tried sync.

```
cp /usr/share/zoneinfo/UTC /etc/localtime
```

So to reproduce, set the system timezone to 'Australia/Sydney' (aka 'AEST')

## Investigation

The following what I found while trying to get it to work. It seems to have been introduced in #5729, since most of the following snippets / functions were added in that PR.

From the traceback:

[make-current-db-time](https://github.com/metabase/metabase/blob/5688c060de14b88481fc2f90df6b6f3baf35feb3/src/metabase/driver.clj#L206)

```clojure
(defn make-current-db-time-fn
  "Takes a clj-time date formatter `DATE-FORMATTER` and a native query
  for the current time. Returns a function that executes the query and
  parses the date returned preserving it's timezone"
  [date-formatter native-query]
  (fn [driver database]
    (some->> (execute-query driver {:database database, :native {:query native-query}})
             :rows
             ffirst
(tformat/parse date-formatter))))
```

[Which is called here during driver initialisation (it seems)](https://github.com/metabase/metabase/blob/4e41782cbfaad70e096a2031bca246f62d136d29/src/metabase/driver/postgres.clj#L236)

```clojure
:current-db-time (driver/make-current-db-time-fn pg-date-formatter pg-db-time-query)}
```

[These arguments are defined just above here](https://github.com/metabase/metabase/blob/4e41782cbfaad70e096a2031bca246f62d136d29/src/metabase/driver/postgres.clj#L189)

```clojure
(def ^:private pg-date-formatter (driver/create-db-time-formatter "yyyy-MM-dd HH:mm:ss.SSS zzz"))
(def ^:private pg-db-time-query "select to_char(current_timestamp, 'YYYY-MM-DD HH24:MI:SS.MS TZ')")
```

This date-formatter seems to be using [Joda time](http://www.joda.org/joda-time/key_format.html) syntax.

[This seems to be the test for the Postgres driver timezone handling, UTC only](https://github.com/metabase/metabase/blob/4e41782cbfaad70e096a2031bca246f62d136d29/test/metabase/driver/postgres_test.clj#L277).

```clojure
(expect-with-engine :postgres
  "UTC"
(tu/db-timezone-id))
```

So a fix might be as simple as changing the pattern to add an extra `z`, e.g.: `yyyy-MM-dd HH:mm:ss.SSS zzzz`. Or it might be much more complicated, #5729 looks pretty hectic.

**Comment by senior** [comment]
@lindsay-stevens Great catch on this bug! I should be able to dig in next week.

**Comment by senior** [comment]
@lindsay-stevens I think there are going to be two parts to fixing this issue. First of all, failing to parse the timestamp shouldn't cause your sync process to fail. I just pushed up a commit that will guard against that (it's in PR #5895). Once that gets merged, it should fix your issue. You also shouldn't have any problems caused by this as it's not being used yet. We hope to eventually have some timezone debugging pages that will use it, but they don't exist yet.

The second part of this issue is that Joda Time doesn't recognize the AEST timezone name. Joda Time is a library that we use for date/time parsing that intends to fix some of the issues with the older built-in JDK date/time parsing library. I have read some about there being ambiguity in those short names and that Joda Time has intentionally not added support for that. For some reason they are supported for US timezones (i.e. CST and PST).

It's our intention to move to the new java.time library once we've dropped support for JDK 7. I tested it out locally and it looks like the new time library has no problems with the AEST timezone. I created #5909 to cover that work.

JDK versions 8+ include a new built-in time library that is the intended successor of both the older JDK date library and Joda Time. I ran some tests locally and it looks like the new library has no problem with the AEST timezone name.   We are intending to move to the new java.time library once we have dropped support for JDK 7.

**Comment by lindsay-stevens** [comment]
@senior Awesome, thanks very much for looking into it. I'll try it after merge / next RC and let you know if there's any hiccups. Seeing as there's new child PR/issues for the problems at hand I'm OK with closing this ticket, unless you want to wait for my feedback.

**Comment by senior** [comment]
@lindsay-stevens Closing it is fine with me. Would be great to have you try it out too. Thanks for reporting the issue!

**Comment by mazameli** [comment]
@lindsay-stevens just a heads up that [we've just shipped version 0.26](http://www.metabase.com/blog/Metabase-0.26) if you'd like to try this out now. Thanks!

**Comment by lindsay-stevens** [comment]
@mazameli thanks! Is it possible to add a jar to the releases page? Or is there a way to get them from CI builds?
